### ISSUE TITLE
RFC: Context Menu API

### ISSUE THREAD

**Issue by tsl0922** [mpv-player/mpv#13677]
### Expected behavior of the wanted feature

I'd like to upstream the C plugin part of [mpv-menu-plugin](https://github.com/tsl0922/mpv-menu-plugin) to mpv.

See also: https://github.com/mpv-player/mpv/issues/5500#issuecomment-1849611413

#### Things need to be do in mpv:

- Add a new property or command to update the menu
- Add a new command to show the context menu

The reference implementation for win32 can be found in [menu.c](https://github.com/tsl0922/mpv-menu-plugin/blob/main/src/menu.c).

#### Benefits

- Less core menu logic in C, support for macOS/Linux can be added later
- menu conf syntax can be implemented with scripts

[dyn_menu.lua](https://github.com/tsl0922/mpv-menu-plugin/blob/main/src/lua/dyn_menu.lua) is a lua script with support for reading menu from `input.conf`, and support dynamic menu items like tracks/chapters/...

#### Additional information

menu data property structure used by mpv-menu-plugin:
```
MPV_FORMAT_NODE_ARRAY
  MPV_FORMAT_NODE_MAP (menu item)
     "type"           MPV_FORMAT_STRING     (supported type: separator, submenu)
     "title"          MPV_FORMAT_STRING     (required if type is not separator)
     "cmd"            MPV_FORMAT_STRING     (optional)
     "state"          MPV_FORMAT_NODE_ARRAY[MPV_FORMAT_STRING] (supported state: checked, disabled, hidden)
     "submenu"        MPV_FORMAT_NODE_ARRAY[menu item]         (required if type is submenu)
```

------------

If this RFC is not accepted, I would request #13163 to be resolved (to make mpv-menu-plugin possible on macOS).

**Comment by sfan5** [comment]
> support for macOS/Linux can be added later

This raises the question: which GUI toolkit do you want mpv to import to implement the context menu?
The question is pointless however because mpv has always avoided GUI toolkits (see e.g. Wayland CSD) and I don't forsee that changing.

Effectively such a feature would need a libass-based fallback anyway. That would also fix the concern of adding a feature (with high visiblity) that only works on on or two platforms.

**Comment by tsl0922** [comment]
> Which GUI toolkit do you want mpv to import to implement the context menu?

I prefer the native API provided by each platform (for native UI experience), for example win32 [Menus](https://learn.microsoft.com/en-us/windows/win32/menurc/menus) is used in [mpv-menu-plugin](https://github.com/tsl0922/mpv-menu-plugin).

**Comment by kasper93** [comment]
> Effectively such a feature would need a libass-based fallback anyway. That would also fix the concern of adding a feature (with high visiblity) that only works one on or two platforms.

For two biggest platforms (macOS and Windows) it can be done using native UI elements. For other platforms like linux, libass fallback can be used or none at all.

**Comment by tsl0922** [comment]
> win32 menus fonts rendering very badly on 2K/4K monitors, whereas libass has no such problems. WinUI should be a better solution on Windows.

Maybe. Win32 menu does scale on 2k/4k, but the default font size and item padding looks bad when the player is in full screen mode, and you can't change it unless creating owner-draw menu, this will introduce complexity.

**Comment by tsl0922** [comment]
> As with the current licensing status quo, any code you want to upstream needs to be relicensed to LGPL.

I'm OK with it, or we may write new code to implement it with WinUI on Windows (if desired).

**Comment by kasper93** [comment]
> There needs to be another property for querying whether native context menu is supported.

Ok, we need to detect whether native menu is working or should we fallback to libass one.
> The script must query this value, and draw the menu on the script side if native menu isn't supported. This also means that the new properties and commands are only useful for native menus.

I don't understand this part? We can use exactly the same interface for libass fallback menu. Backend of handling menu shouldn't change the way it works.

> I'm OK with it, or we may write new code to implement it with WinUI on Windows (if desired).

I don't think using WinUI is feasible. Also I don't think there will be much gain, but in the same time feel free to prove me wrong.

**Comment by kasper93** [comment]
Let's focus on one objective at the time.  There is working win32 backend already implemented, it can be improved / changed, but this issue is about the mpv integration, not WinUI.

**Comment by Dudemanguy** [comment]
This plugin uses the `window-id` property right? How important is that exactly because that will never work on wayland.

**Comment by tsl0922** [comment]
> This plugin uses the `window-id` property right? How important is that exactly because that will never work on wayland.

The window handle is used to check whether the mouse is clicked on window, it's not required to context menu.

In fact, a C plugin always run in it's own thread, this make it complicated to implement GUI code, because GUI code typically need to run in the GUI thread (may or may not required, ex: main thread is required on macOS). This is a reason why I want to merge the code to mpv core, it's easier because we can write code in the GUI thread directly.

**Comment by kasper93** [comment]
It has been merged as #13700
